sklearn/manifold/_t_sne.py

```diff
@@ -715,15 +739,23 @@ def _fit(self, X, skip_num_points=0):
                     print("[t-SNE] Computing pairwise distances...")
 
                 if self.metric == "euclidean":
+                    # Euclidean is squared here, rather than using **= 2,
+                    # because euclidean_distances already calculates
+                    # squared distances, and returns np.sqrt(dist) for
+                    # squared=False.
                     distances = pairwise_distances(X, metric=self.metric,
                                                    squared=True)
                 else:
+                    # Euclidean has separate call as it's slower for n_jobs>1
                     distances = pairwise_distances(X, metric=self.metric,
                                                    n_jobs=self.n_jobs)
 
-                if np.any(distances < 0):
-                    raise ValueError("All distances should be positive, the "
-                                     "metric given is not correct")
+            if np.any(distances < 0):
+                raise ValueError("All distances should be positive, the "
+                                 "metric given is not correct")
+
+            if self.metric != "euclidean" and self.square_distance is True:
```

Comment:
Does this also work in interpreters that aren't CPython? I thought that True being a singleton was a CPython implementation detail

Reply:
This might not be thorough enough, but a quick search yields:
* IronPython: https://ironpython-test.readthedocs.io/en/latest/library/constants.html
* Jython: https://www.jython.org/jython-old-sites/docs/library/stdtypes.html#boolean-values
* PyPy: I couldn't find reference to bools in https://doc.pypy.org/en/latest/cpython_differences.html#object-identity-of-primitive-values-is-and-id, but I did attempt to quickly test it here: https://pypyjs.org/ See below:

![Screenshot from 2020-07-10 16-17-05](https://user-images.githubusercontent.com/16181459/87210097-37497f00-c2c9-11ea-8239-dbe5a7895b87.png)